File: src/constants/resourcePaths.ts
export const RESOURCE_DB_FILES = {
  items: 'items_db.json',
  skills: 'skills_db.json',
  monsters: 'monsters_db.json',
  eventEncounters: 'event_encounters.json',
  eventDetail: 'event_detail.json',
  combatEncounters: 'combat_encounters.json',
  cardIdMapping: 'card_id_mapping.json',
} as const;

export const RESOURCE_DIR_ALIASES: Array<{ canonical: string; legacy: string }> = [
  { canonical: 'assets/items', legacy: 'images' },
  { canonical: 'assets/skills', legacy: 'images/skill' },
  { canonical: 'assets/monsters/characters', legacy: 'images_monster_char' },
  { canonical: 'assets/monsters/backgrounds', legacy: 'images_monster_bg' },
  { canonical: 'assets/gui', legacy: 'images_GUI' },
  { canonical: 'assets/events/characters', legacy: 'EncEvent_CHAR' },
  { canonical: 'assets/events/backgrounds', legacy: 'EncEvent_BG' },
  { canonical: 'assets/events/icons', legacy: 'EncEvent_Icons' },
  { canonical: 'assets/sponsor', legacy: 'sponsor' },
];

export const normalizeResourcePath = (rawPath: string): string => {
  return rawPath.replace(/^resources\//, '').replace(/^\//, '');
};

const mapPrefix = (value: string, fromPrefix: string, toPrefix: string): string | null => {
  if (value === fromPrefix) return toPrefix;
  if (value.startsWith(`${fromPrefix}/`)) {
    return `${toPrefix}/${value.slice(fromPrefix.length + 1)}`;
  }
  return null;
};

export const buildResourceCandidates = (rawPath: string): string[] => {
  const normalized = normalizeResourcePath(rawPath);
  const candidates = new Set<string>([normalized]);

  for (const alias of RESOURCE_DIR_ALIASES) {
    const fromCanonical = mapPrefix(normalized, alias.canonical, alias.legacy);
    if (fromCanonical) candidates.add(fromCanonical);

    const fromLegacy = mapPrefix(normalized, alias.legacy, alias.canonical);
    if (fromLegacy) candidates.add(fromLegacy);
  }

  return [...candidates];
};

File: src-tauri/src/data_management/resource_paths.rs
use std::path::{Path, PathBuf};

pub const RESOURCE_DB_FILES: &[&str] = &[
    "items_db.json",
    "skills_db.json",
    "monsters_db.json",
    "event_encounters.json",
    "event_detail.json",
    "combat_encounters.json",
    "card_id_mapping.json",
];

pub const RESOURCE_DIR_ALIASES: &[(&str, &str)] = &[
    ("assets/items", "images"),
    ("assets/skills", "images/skill"),
    ("assets/monsters/characters", "images_monster_char"),
    ("assets/monsters/backgrounds", "images_monster_bg"),
    ("assets/gui", "images_GUI"),
    ("assets/events/characters", "EncEvent_CHAR"),
    ("assets/events/backgrounds", "EncEvent_BG"),
    ("assets/events/icons", "EncEvent_Icons"),
    ("assets/sponsor", "sponsor"),
];

fn map_prefix(value: &str, from_prefix: &str, to_prefix: &str) -> Option<String> {
    if value == from_prefix {
        return Some(to_prefix.to_string());
    }

    value
        .strip_prefix(&(from_prefix.to_string() + "/"))
        .map(|rest| format!("{to_prefix}/{rest}"))
}

pub fn build_resource_candidates(raw_path: &str) -> Vec<String> {
    let normalized = raw_path
        .trim_start_matches("resources/")
        .trim_start_matches('/');

    let mut output = vec![normalized.to_string()];

    for (canonical, legacy) in RESOURCE_DIR_ALIASES {
        if let Some(mapped) = map_prefix(normalized, canonical, legacy) {
            if !output.contains(&mapped) {
                output.push(mapped);
            }
        }

        if let Some(mapped) = map_prefix(normalized, legacy, canonical) {
            if !output.contains(&mapped) {
                output.push(mapped);
            }
        }
    }

    output
}

pub fn resolve_existing_resource(resources_dir: &Path, raw_path: &str) -> Option<PathBuf> {
    for candidate in build_resource_candidates(raw_path) {
        let full = resources_dir.join(&candidate);
        if full.exists() {
            return Some(full);
        }
    }
    None
}
